File: jcms/jcms-core/src/main/java/com/ysh/jcms/svc/report/CmsSetUrcbResult.java
package com.ysh.jcms.svc.report;

import com.ysh.jcms.core.CmsType;
import com.ysh.jcms.data.common.CmsServiceError;
import com.ysh.jcms.data.scalar.CmsBoolean;
import java.util.Arrays;
import java.util.List;

/**
 * SetURCBResult ::= SEQUENCE {
 *     error       [0] IMPLICIT ServiceError OPTIONAL,
 *     rptID       [1] IMPLICIT ServiceError OPTIONAL,
 *     rptEna      [2] IMPLICIT ServiceError OPTIONAL,
 *     datSet      [3] IMPLICIT ServiceError OPTIONAL,
 *     optFlds     [5] IMPLICIT ServiceError OPTIONAL,
 *     bufTm       [6] IMPLICIT ServiceError OPTIONAL,
 *     trgOps      [8] IMPLICIT ServiceError OPTIONAL,
 *     intgPd      [9] IMPLICIT ServiceError OPTIONAL,
 *     gi          [10] IMPLICIT ServiceError OPTIONAL,
 *     resv        [13] IMPLICIT ServiceError OPTIONAL
 * }  —  8.7.5
 */
public class CmsSetUrcbResult extends CmsType {

    public CmsBoolean        errorPresent;
    public CmsServiceError   error;            /* OPTIONAL */
    public CmsBoolean        rptIdErrPresent;
    public CmsServiceError   rptIdErr;         /* OPTIONAL */
    public CmsBoolean        rptEnaErrPresent;
    public CmsServiceError   rptEnaErr;        /* OPTIONAL */
    public CmsBoolean        datSetErrPresent;
    public CmsServiceError   datSetErr;        /* OPTIONAL */
    public CmsBoolean        optFldsErrPresent;
    public CmsServiceError   optFldsErr;       /* OPTIONAL */
    public CmsBoolean        bufTmErrPresent;
    public CmsServiceError   bufTmErr;         /* OPTIONAL */
    public CmsBoolean        trgOpsErrPresent;
    public CmsServiceError   trgOpsErr;        /* OPTIONAL */
    public CmsBoolean        intgPdErrPresent;
    public CmsServiceError   intgPdErr;        /* OPTIONAL */
    public CmsBoolean        giErrPresent;
    public CmsServiceError   giErr;            /* OPTIONAL */
    public CmsBoolean        resvErrPresent;
    public CmsServiceError   resvErr;          /* OPTIONAL */

    public CmsSetUrcbResult() {
        this.errorPresent      = new CmsBoolean();
        this.error             = new CmsServiceError();
        this.rptIdErrPresent   = new CmsBoolean();
        this.rptIdErr          = new CmsServiceError();
        this.rptEnaErrPresent  = new CmsBoolean();
        this.rptEnaErr         = new CmsServiceError();
        this.datSetErrPresent  = new CmsBoolean();
        this.datSetErr         = new CmsServiceError();
        this.optFldsErrPresent = new CmsBoolean();
        this.optFldsErr        = new CmsServiceError();
        this.bufTmErrPresent   = new CmsBoolean();
        this.bufTmErr          = new CmsServiceError();
        this.trgOpsErrPresent  = new CmsBoolean();
        this.trgOpsErr         = new CmsServiceError();
        this.intgPdErrPresent  = new CmsBoolean();
        this.intgPdErr         = new CmsServiceError();
        this.giErrPresent      = new CmsBoolean();
        this.giErr             = new CmsServiceError();
        this.resvErrPresent    = new CmsBoolean();
        this.resvErr           = new CmsServiceError();
    }
    
    // -- chain setters --
    public CmsSetUrcbResult errorPresent(boolean v) { this.errorPresent.value(v); return this; }
    public CmsSetUrcbResult error(int v) { this.error.value(v); return this; }
    public CmsSetUrcbResult rptIdErrPresent(boolean v) { this.rptIdErrPresent.value(v); return this; }
    public CmsSetUrcbResult rptIdErr(int v) { this.rptIdErr.value(v); return this; }
    public CmsSetUrcbResult rptEnaErrPresent(boolean v) { this.rptEnaErrPresent.value(v); return this; }
    public CmsSetUrcbResult rptEnaErr(int v) { this.rptEnaErr.value(v); return this; }
    public CmsSetUrcbResult datSetErrPresent(boolean v) { this.datSetErrPresent.value(v); return this; }
    public CmsSetUrcbResult datSetErr(int v) { this.datSetErr.value(v); return this; }
    public CmsSetUrcbResult optFldsErrPresent(boolean v) { this.optFldsErrPresent.value(v); return this; }
    public CmsSetUrcbResult optFldsErr(int v) { this.optFldsErr.value(v); return this; }
    public CmsSetUrcbResult bufTmErrPresent(boolean v) { this.bufTmErrPresent.value(v); return this; }
    public CmsSetUrcbResult bufTmErr(int v) { this.bufTmErr.value(v); return this; }
    public CmsSetUrcbResult trgOpsErrPresent(boolean v) { this.trgOpsErrPresent.value(v); return this; }
    public CmsSetUrcbResult trgOpsErr(int v) { this.trgOpsErr.value(v); return this; }
    public CmsSetUrcbResult intgPdErrPresent(boolean v) { this.intgPdErrPresent.value(v); return this; }
    public CmsSetUrcbResult intgPdErr(int v) { this.intgPdErr.value(v); return this; }
    public CmsSetUrcbResult giErrPresent(boolean v) { this.giErrPresent.value(v); return this; }
    public CmsSetUrcbResult giErr(int v) { this.giErr.value(v); return this; }
    public CmsSetUrcbResult resvErrPresent(boolean v) { this.resvErrPresent.value(v); return this; }
    public CmsSetUrcbResult resvErr(int v) { this.resvErr.value(v); return this; }
    @Override
    public List<? extends CmsType> children() {
        return Arrays.asList(errorPresent, error,
            rptIdErrPresent, rptIdErr,
            rptEnaErrPresent, rptEnaErr,
            datSetErrPresent, datSetErr,
            optFldsErrPresent, optFldsErr,
            bufTmErrPresent, bufTmErr,
            trgOpsErrPresent, trgOpsErr,
            intgPdErrPresent, intgPdErr,
            giErrPresent, giErr,
            resvErrPresent, resvErr);
    }
}

File: ccms/src/svc/report/cms_set_urcb_result.c
#include "svc/report/cms_set_urcb_result.h"
#include "data/scalar/cms_boolean.h"
#include "data/common/cms_service_error.h"
#include "per/cms_sequence.h"

int cms_set_urcb_result_encode_stream(per_stream_t *s, const cms_set_urcb_result_t *v) {
    if (!v)
        return CMS_ERR;
    int err;

    /* 0. OPTIONAL bitmap (10 fields: error, rptID, rptEna, datSet, optFlds, bufTm, trgOps, intgPd, gi, resv) */
    bool opt_present[10] = {(v->error_present && v->error_present->value) && v->error,
                            (v->rpt_id_err_present && v->rpt_id_err_present->value) && v->rpt_id_err,
                            (v->rpt_ena_err_present && v->rpt_ena_err_present->value) && v->rpt_ena_err,
                            (v->dat_set_err_present && v->dat_set_err_present->value) && v->dat_set_err,
                            (v->opt_flds_err_present && v->opt_flds_err_present->value) && v->opt_flds_err,
                            (v->buf_tm_err_present && v->buf_tm_err_present->value) && v->buf_tm_err,
                            (v->trg_ops_err_present && v->trg_ops_err_present->value) && v->trg_ops_err,
                            (v->intg_pd_err_present && v->intg_pd_err_present->value) && v->intg_pd_err,
                            (v->gi_err_present && v->gi_err_present->value) && v->gi_err,
                            (v->resv_err_present && v->resv_err_present->value) && v->resv_err};
    err = (int) per_encode_optional_bitmap(s, opt_present, 10);
    if (err)
        return err;

    /* 1. error — ServiceError OPTIONAL (bitmap[0]) */
    if (opt_present[0]) {
        err = cms_service_error_encode_stream(s, v->error);
        if (err)
            return err;
    }

    /* 2. rptIdErr — ServiceError OPTIONAL (bitmap[1]) */
    if (opt_present[1]) {
        err = cms_service_error_encode_stream(s, v->rpt_id_err);
        if (err)
            return err;
    }

    /* 3. rptEnaErr — ServiceError OPTIONAL (bitmap[2]) */
    if (opt_present[2]) {
        err = cms_service_error_encode_stream(s, v->rpt_ena_err);
        if (err)
            return err;
    }

    /* 4. datSetErr — ServiceError OPTIONAL (bitmap[3]) */
    if (opt_present[3]) {
        err = cms_service_error_encode_stream(s, v->dat_set_err);
        if (err)
            return err;
    }

    /* 5. optFldsErr — ServiceError OPTIONAL (bitmap[4]) */
    if (opt_present[4]) {
        err = cms_service_error_encode_stream(s, v->opt_flds_err);
        if (err)
            return err;
    }

    /* 6. bufTmErr — ServiceError OPTIONAL (bitmap[5]) */
    if (opt_present[5]) {
        err = cms_service_error_encode_stream(s, v->buf_tm_err);
        if (err)
            return err;
    }

    /* 7. trgOpsErr — ServiceError OPTIONAL (bitmap[6]) */
    if (opt_present[6]) {
        err = cms_service_error_encode_stream(s, v->trg_ops_err);
        if (err)
            return err;
    }

    /* 8. intgPdErr — ServiceError OPTIONAL (bitmap[7]) */
    if (opt_present[7]) {
        err = cms_service_error_encode_stream(s, v->intg_pd_err);
        if (err)
            return err;
    }

    /* 9. giErr — ServiceError OPTIONAL (bitmap[8]) */
    if (opt_present[8]) {
        err = cms_service_error_encode_stream(s, v->gi_err);
        if (err)
            return err;
    }

    /* 10. resvErr — ServiceError OPTIONAL (bitmap[9]) */
    if (opt_present[9]) {
        err = cms_service_error_encode_stream(s, v->resv_err);
        if (err)
            return err;
    }

    return CMS_OK;
}

int cms_set_urcb_result_decode_stream(per_stream_t *s, void *ptr) {
    cms_set_urcb_result_t *v = (cms_set_urcb_result_t *) ptr;
    int err;

    /* 0. OPTIONAL bitmap (10 fields) */
    bool opt_present[10];
    err = (int) per_decode_optional_bitmap(s, opt_present, 10);
    if (err)
        return err;
    if (v) {
        if (v->error_present)
            v->error_present->value = opt_present[0];
        if (v->rpt_id_err_present)
            v->rpt_id_err_present->value = opt_present[1];
        if (v->rpt_ena_err_present)
            v->rpt_ena_err_present->value = opt_present[2];
        if (v->dat_set_err_present)
            v->dat_set_err_present->value = opt_present[3];
        if (v->opt_flds_err_present)
            v->opt_flds_err_present->value = opt_present[4];
        if (v->buf_tm_err_present)
            v->buf_tm_err_present->value = opt_present[5];
        if (v->trg_ops_err_present)
            v->trg_ops_err_present->value = opt_present[6];
        if (v->intg_pd_err_present)
            v->intg_pd_err_present->value = opt_present[7];
        if (v->gi_err_present)
            v->gi_err_present->value = opt_present[8];
        if (v->resv_err_present)
            v->resv_err_present->value = opt_present[9];
    }

    /* 1. error OPTIONAL */ if (opt_present[0]) {
        if (v && !v->error)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->error : NULL);
        if (err)
            return err;
    }
    /* 2. rptIdErr OPTIONAL */ if (opt_present[1]) {
        if (v && !v->rpt_id_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->rpt_id_err : NULL);
        if (err)
            return err;
    }
    /* 3. rptEnaErr OPTIONAL */ if (opt_present[2]) {
        if (v && !v->rpt_ena_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->rpt_ena_err : NULL);
        if (err)
            return err;
    }
    /* 4. datSetErr OPTIONAL */ if (opt_present[3]) {
        if (v && !v->dat_set_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->dat_set_err : NULL);
        if (err)
            return err;
    }
    /* 5. optFldsErr OPTIONAL */ if (opt_present[4]) {
        if (v && !v->opt_flds_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->opt_flds_err : NULL);
        if (err)
            return err;
    }
    /* 6. bufTmErr OPTIONAL */ if (opt_present[5]) {
        if (v && !v->buf_tm_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->buf_tm_err : NULL);
        if (err)
            return err;
    }
    /* 7. trgOpsErr OPTIONAL */ if (opt_present[6]) {
        if (v && !v->trg_ops_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->trg_ops_err : NULL);
        if (err)
            return err;
    }
    /* 8. intgPdErr OPTIONAL */ if (opt_present[7]) {
        if (v && !v->intg_pd_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->intg_pd_err : NULL);
        if (err)
            return err;
    }
    /* 9. giErr OPTIONAL */ if (opt_present[8]) {
        if (v && !v->gi_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->gi_err : NULL);
        if (err)
            return err;
    }
    /* 10. resvErr OPTIONAL */ if (opt_present[9]) {
        if (v && !v->resv_err)
            return CMS_ERR;
        err = cms_service_error_decode_stream(s, v ? v->resv_err : NULL);
        if (err)
            return err;
    }

    return CMS_OK;
}
